File: app/src/main/java/avatar/com/mymvptestone/java/mvp/base/BasePresenter.java
package avatar.com.mymvptestone.java.mvp.base;

import java.lang.ref.WeakReference;

/**
 * $desc$
 *
 * @author Frank
 * email xiong.fu@avatarmind.com
 * created 19-4-4 下午1:26.
 */
public abstract class BasePresenter<M extends Model, V extends View> implements Presenter<V,M> {

    protected M model;

    protected WeakReference<V> wrfView;

    @Override
    public void registerModel(M model) {
        this.model = model;
    }

    @Override
    public void registerView(V view) {
        this.wrfView = new WeakReference<V>(view);
    }

    @Override
    public V getView() {
        return wrfView == null? null : wrfView.get();
    }

    @Override
    public void destroy() {
        if (wrfView != null){
            wrfView.clear();
            wrfView = null;
        }
        onViewDestroy();
    }

    public abstract void onViewDestroy();
}

File: app/src/main/java/avatar/com/mymvptestone/kotlin/mvp/base/BasePresenter.kt
package avatar.com.mymvptestone.kotlin.mvp.base

import java.lang.ref.WeakReference

/**
 * $desc$
 * @author Frank
 * email xiong.fu@avatarmind.com
 * created 19-4-4 下午3:54.
 */
abstract class BasePresenter<M: Model, V: View> : Presenter<M,V>{
    var model: M? = null
    var view: WeakReference<V>? = null

    override fun registerModel(model: M?) {
        this.model = model
    }

    override fun registerView(view: V?) {
        this.view = WeakReference<V>(view)
    }

    override fun getView(): V? {
        return view?.get()
    }

    override fun destroy() {
        view?.clear()
        onViewDestroy()
    }

    abstract fun onViewDestroy()
}